### ISSUE TITLE
Massive FPS drop around some polygons

### ISSUE THREAD

**Issue by RicoElectrico** [mapsme/omim#1238]
MAPS.ME doesn't seem to like these polygons (namely water and the forest north of it). Tested on the latest (2016-01-05) build. Map version is the latest available in-app.
Video: https://gfycat.com/AlarmedCarefulDragon
Location: http://www.openstreetmap.org/#map=15/54.5436/18.5293

**Comment by rokuz** [comment]
Hi, thanks for reporting. I've got a couple of questions:
1) On which device do you run the app?
2) How to you measure FPS? Why do you think that FPS dropped? In the video I see that some areas are loaded with delay. Is it the described problem?

**Comment by RicoElectrico** [comment]
1) LG L90
2) I didn't, it's visible to the naked eye ;) Normally when elements load up they don't block anything else. The performance drop is obvious when you do a kinetic scroll - the animation gets considerably choppier when it approaches that forest or port canal. It seems to get worse on higher zoom levels where any touch manipulation on the map gets very very sluggish. 
Sadly gfycat is limited only to 15 seconds and it's not evident on the video, but otherwise performance is much better. Compare with this video which I made earlier: http://gfycat.com/PalePerfectBetafish

**Comment by rokuz** [comment]
Do you mean that you had got a version with 3D buildings where kinetic scroll worked fine? When did you capture the second video?

**Comment by rokuz** [comment]
Let's check I get the problem correctly. When we are above large polygons (like forests or seas) or on high zoom levels, the kinetic scroll becomes sluggish. If we are in a city (where a lot of small polygons), the kinetic scroll is fine. The problem have become recently, when you recorded the second video the kinetic scroll worked without freezes anywhere?

**Comment by RicoElectrico** [comment]
Damn. I thought it's simple to explain :P It has nothing to do with kinetic scroll itself, it just happens that's a good way to demonstrate how FPS drops. So let me rephrase: Whenever I get close to these specific polygons, MAPS.ME gets very slow, touch reactions and other animations get very stuttering. For now it seems that it's not directly related to the polygon size, there are much bigger polygons that draw just fine and smoothly.
Did you download Poland map and reproduce the bug for yourself in that area?

**Comment by rokuz** [comment]
Thank you for detailed explanation. Now it seems clear :) I had to ask to get the problem precisely :) I've downloaded Poland map. On Nexus 6 I couldn't reproduce it, on weak device (Samsung s4 mini) I hardly found the area where the problem had reproduced. So, yes, the bug exists.
On the face of it I'm not sure the bug is in MAPS.ME code. I suppose it can be caused by GPU (which is inside Samsung s4 mini and your LG L90) Adreno 305. Anyway we're going to investigate the problem precisely. Please keep us in touch if you find new details.

**Comment by RicoElectrico** [comment]
I wanted to check it elsewhere and I downloaded the latest build into Xperia J (Android 4.2), but here MAPS.ME is outright unusable even at the world level. There are numerous graphical glitches, like flashing, inverted colors, redraws (the infamous hall of mirrors effect). It's totally broken because you can't read anything on the map.
But as it's not my phone, I couldn't investigate further to give you more details.

BTW, is there any feature in MAPS.ME to dump some debug info?

**Comment by utack** [comment]
>  There are numerous graphical glitches, like flashing, inverted colors, redraws
Probably just Qualcomms GPU driver mess.
Android 4.2 sounds like it has an ancient GPU driver, they were even worse back then

**Comment by RicoElectrico** [comment]
Google Maps works fine here. Well, it works just fine even on an ancient 2.3.5 HTC Wildfire S, 3D buildings included. They clearly use some alien technology, right? :stuck_out_tongue_winking_eye:

**Comment by rokuz** [comment]
You can use the expression LOG(LDEBUG, (smth)) to write everything you want from code to the log. It will be in logcat in case of Android.

**Comment by RicoElectrico** [comment]
@rokuz @utack I forgot to mention, this is a regression. It worked just fine, certainly before the recent release which introduced "don't stop loading data when zooming" :wink:  (or maybe 3D buildings). Where do I download the old releases?

**Comment by rokuz** [comment]
http://maps.me/apk/

**Comment by RicoElectrico** [comment]
The issue seems to be gone on the latest build.

**Comment by rokuz** [comment]
Yeap, we've fixed it :)
